### DISCUSSION TITLE
Listening Queue

### DISCUSSION THREAD

**Discussion by taehyun517** [ANLAB-KAIST/KENSv3#203] [Project Q&A 2. Transport Layer (TCP)]
At the MultipleInterface2 test, the backlog of the MultipleInterface2 is set to 1 and two ip addresses are bind to a socket. I have several questions about this test and the solution's behavior.
1. Even though there is only one socket, why does it call listen() 2 times?
<img width="359" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/85110272/915fde54-f80f-454f-9f20-733229d76fa1">
2. In the test case, there are two SYNs coming almost simultaneously. However, how is it possible to put those two connections to a queue where the backlog is 1? 
<img width="959" alt="스크린샷 2024-04-11 205525" src="https://github.com/ANLAB-KAIST/KENSv3/assets/85110272/f2fc7989-5de9-4362-9760-acbdff4a97dd">

In my implementation, one SYN gets retransmitted. I'm confused about the what the queue should do. So, is SYNACK->push to queue
and ACK->pop it out?

**Comment by Pusnow** [comment]
> 1. Even though there is only one socket, why does it call listen() 2 times?

The `listen()`s are called in the different processes (which has different `pid`s). So you have to separate data structures for each process. The `fd`s are also different even though the numbers are same (3) because they have different pids.

> 2. In the test case, there are two SYNs coming almost simultaneously. However, how is it possible to put those two connections to a queue where the backlog is 1?

They have two different IP addresses and sockets, so each can handle each handshake.

> In my implementation, one SYN gets retransmitted. I'm confused about the what the queue should do. So, is SYNACK->push to queue
and ACK->pop it out?

Again, if each process (or pid) has separate data structure, it would succeed.
